lib/format_parser.rb

```diff
@@ -8,21 +8,34 @@ module FormatParser
   require_relative 'remote_io'
   require_relative 'io_constraint'
   require_relative 'care'
-  require_relative 'parsers/dsl'
 
   PARSER_MUX = Mutex.new
 
-  def self.register_parser_constructor(object_responding_to_new)
+  def self.register_parser(callable_or_responding_to_new, formats:, natures:)
+    parser_provided_formats = Array(formats)
+    parser_provided_natures = Array(natures)
     PARSER_MUX.synchronize do
-      @parsers ||= []
-      @parsers << object_responding_to_new
-      # Gathering natures and formats from parsers. An instance has to be created.
-      parser = object_responding_to_new.new
-      @natures ||= Set.new
-      # NOTE: merge method for sets modify the instance.
-      @natures.merge(parser.natures)
-      @formats ||= Set.new
-      @formats.merge(parser.formats)
+      @parsers ||= Set.new
+      @parsers << callable_or_responding_to_new
+      @natures_per_parser ||= {}
+      parser_provided_natures.each do |provided_nature|
+        @natures_per_parser[provided_nature] ||= Set.new
+        @natures_per_parser[provided_nature] << callable_or_responding_to_new
+      end
+      @formats_per_parser ||= {}
+      parser_provided_formats.each do |provided_format|
+        @formats_per_parser[provided_format] ||= Set.new
+        @formats_per_parser[provided_format] << callable_or_responding_to_new
+      end
+    end
+  end
+
+  def self.deregister_parser(callable_or_responding_to_new)
+    # Used only in tests
+    PARSER_MUX.synchronize do
+      (@parsers || []).delete(callable_or_responding_to_new)
+      (@natures_per_parser || {}).values.map { |e| e.delete(callable_or_responding_to_new) }
```

Comment:
I'd use `&` operator to be nil-safe.

Reply:
How would `&` help here? `&` is a bitwise AND. And (and!) everything in `.values` here must be a Set because the only place where we assign to it is in the mirror method to register. Or am I missing something?

Reply:
Oh dude, I didn't express myself correctly. What I meant something like 

`@parsers&.delete(...)`

For sure it's not the `&` operator, but yeah, this syntax can help to avoid having to create an empty-default-ish object among those lines.

Reply:
Ah good one! I still have to get used to that trick.